src/material_library.cpp

```diff
@@ -0,0 +1,342 @@
+  from_json(filename);
+
+}
+void pyne::MaterialLibrary::from_json(const std::string& filename) {
+  if (!pyne::file_exists(filename))
+    throw pyne::FileNotFound(filename);
+  std::string s;
+  std::ifstream f (filename.c_str(), std::ios::in | std::ios::binary);
+  f.seekg(0, std::ios::end);
+  s.resize(f.tellg());
+  f.seekg(0, std::ios::beg);
+  f.read(&s[0], s.size());
+  f.close();
+  Json::Reader reader;
+  Json::Value json;
+  reader.parse(s, json);
+  load_json(json);
+}
+
+void pyne::MaterialLibrary::write_json(char * filename) {
+  std::string fname (filename);
+  write_json(fname);
+}
+
+void pyne::MaterialLibrary::write_json(const std::string& filename) {
+  Json::Value json = dump_json();
+  Json::StyledWriter writer;
+  std::string s = writer.write(json);
+  std::ofstream f;
+  f.open(filename.c_str(), std::ios_base::trunc);
+  f << s << "\n";
+  f.close();
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat) {
+  // if exists, get the material name from metadata make one instead
+  std::string mat_name;
+  std::pair<pyne::matname_set::iterator, bool> mat_insert;
+  if (mat.metadata.isMember("name")) {
+    mat_name = mat.metadata["name"].asString();
+    mat_insert = matlist.insert(mat.metadata["name"].asString());
+  } else {
+    // form a mat name as 'mX'
+    int mat_number = name_order.size() +1;
+    mat_name = "m" + std::to_string(mat_number);
+    mat.metadata["name"] = mat_name;
+    mat_insert = matlist.insert(mat.metadata["name"].asString());
+
+    // if 'mX' name is already in the Library rename the material
+    while (!mat_insert.second) {
+      mat_number += name_order.size();
+      mat_name = "m" + std::to_string(mat_number);
+      mat.metadata["name"] = mat_name;
+      mat_insert = matlist.insert(mat.metadata["name"].asString());
+    }
+  }
+  if (mat_insert.second) {
+    append_to_nuclist(mat);
+    material_library[mat_name] = new Material(mat);
+    material_library[mat_name]->metadata["mat_number"] = int(name_order.size() +1);
+    
+    name_order.push_back(mat_name);
+  }
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat, char* mat_name) {
+  std::string material_name(mat_name);
+  add_material(mat, material_name);
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat,
+                                         const std::string& mat_name) {
+  mat.metadata["name"] = mat_name;
+  add_material(mat);
+}
+
+void pyne::MaterialLibrary::del_material(pyne::Material mat) {
+  if (mat.metadata.isMember("name")) {
+    std::string mat_name = mat.metadata["name"].asString();
+    del_material(mat_name);
+  }
+}
+
+void pyne::MaterialLibrary::del_material(const std::string& mat_name) {
+  material_library.erase(mat_name);
+  matlist.erase(mat_name);
+  for ( auto name = name_order.begin(); name != name_order.end(); name++) {
+    if (*name == mat_name) {
+      name_order.erase(name);
+      return;
+    }
+  }
+}
+
+pyne::Material pyne::MaterialLibrary::get_material(
```

Comment:
This should call `get_material_ptr` (as it will soon be called)

Reply:
this return a copy of the material not the pointer itself.
so I believe the naming is correct.